lib/TPP/Dialect/Tpp/TppOps.cpp

```diff
@@ -268,34 +270,68 @@ void AddOp::getEffects(
 // GemmOp
 //===----------------------------------------------------------------------===//
 
-static bool verifyGemmShape(ShapedType shapedA, ShapedType shapedB,
-                            ShapedType shapedC) {
-  return !(shapedB.getRank() != 2 || shapedC.getRank() != 2 ||
-           shapedA.getRank() != 2);
-}
-
-static bool verifyGemmOperandsDims(ArrayRef<int64_t> shapeA,
-                                   ArrayRef<int64_t> shapeB,
-                                   ArrayRef<int64_t> shapeC) {
-  int64_t m = shapeC[0];
-  int64_t n = shapeC[1];
-  int64_t k = shapeA[1];
-  // Verify C(m, n) = A(m, k) B(k, n)
-  return !(shapeB[0] != k || shapeB[1] != n || shapeA[0] != m);
+template <typename OpTy>
+static LogicalResult
+validateVnniGemmOperand(OpTy operation, ArrayRef<int64_t> shape,
+                        Type elementType, int dimI, int dimJ) {
+  if (shape.size() == 3) {
+    if (!elementType.isBF16()) {
+      return operation->emitOpError() << "operand 1 invalid element type for "
+                                         "VNNI layout expect bf16, but got: "
+                                      << elementType << "\n";
+    }
+    if (shape[shape.size() - 1] !=
+        vnni::utils::getVnniBlockingFactor(elementType)) {
+      return operation->emitOpError() << "operand 1 invalid VNNI layout expect "
+                                         "inner dims to be 2 or 4, but got: "
+                                      << shape[shape.size() - 1] << "\n";
+    }
+    // VNNI layout: [K/VNNI][J][VNNI]
+    if (shape[0] * shape[shape.size() - 1] != dimI || shape[1] != dimJ)
+      return operation->emitOpError("operand 1 fails to verify expected shape");
+    return success();
+  }
+  if (shape.size() != 2) {
+    return operation->emitOpError()
+           << "operand 1 expects rank 2, but got: " << shape.size() << "\n";
+  }
+  if (shape[0] != dimI || shape[1] != dimJ)
```

Comment:
This check could be moved up to avoid the extra indentation on the VNNI case.

Reply:
Sorry, did not get. Where you want to move it? line 289?

Reply:
Yes, make the `shape == 2` the first case, because it's one line check, and the `shape == 3` the remainder, because there are many checks.